exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/ColumnType.kt

```diff
@@ -739,7 +739,21 @@ open class BasicBinaryColumnType : ColumnType<ByteArray>() {
         else -> error("Unexpected value $value of type ${value::class.qualifiedName}")
     }
 
-    override fun nonNullValueToString(value: ByteArray): String = value.toString(Charsets.UTF_8)
+    override fun nonNullValueToString(value: ByteArray): String {
+        return when {
+            currentDialect is H2Dialect && currentDialect.h2Mode == null -> "X'${value.toHexString().lowercase()}'"
+            currentDialect is MariaDBDialect -> toTextString(value)
+            currentDialect is OracleDialect -> "HEXTORAW('${value.toHexString()}')"
+            currentDialect is SQLServerDialect || currentDialect is MysqlDialect -> "0x${value.toHexString()}"
+            currentDialect is PostgreSQLDialect -> "'\\x${value.toHexString().lowercase()}'::bytea"
+            else -> toTextString(value)
+        }
+    }
+
+    private fun toTextString(value: ByteArray) = TextColumnType().nonNullValueToString(value.toString(Charsets.UTF_8))
+
+    @Suppress("MagicNumber")
+    private fun ByteArray.toHexString(): String = joinToString("") { it.toString(16).uppercase().padStart(2, '0') }
```

Comment:
All this and above functionality already exists within the blob (binary) column type. For example, `ExposedBlob` has the same function `toHexString()` that can be extracted and reused.

But if you don't mind the duplicate here, at least `nonNullValueToString()` should use `currentDialect.dataTypeProvider.hexToDb(value.hexString())`. From what I'm seeing, existing `hexToDb()` matches your branches above.

Reply:
TODO fix with [EXPOSED-433](https://youtrack.jetbrains.com/issue/EXPOSED-433) Check all the literals for default/insert/read operations